LibCarla/source/carla/trafficmanager/CollisionStage.cpp

```diff
@@ -166,90 +180,149 @@ namespace CollisionStageConstants {
 
     float inter_vehicle_distance = cg::Math::DistanceSquared(reference_location, other_location);
-    float minimum_inter_vehicle_distance = std::pow(GetBoundingBoxExtention(reference_vehicle) + inter_vehicle_length, 2.0f);
-
-    if (!(!reference_front_wp->CheckJunction() &&
-        cg::Math::Dot(reference_heading, reference_to_other) < 0 &&
-        inter_vehicle_distance > minimum_inter_vehicle_distance) &&
-
-        !(!closest_point->CheckJunction() && junction_look_ahead->CheckJunction() &&
-        reference_vehicle_ptr->GetVelocity().SquaredLength() < 0.1 &&
-        reference_vehicle_ptr->GetTrafficLightState() != carla::rpc::TrafficLightState::Green &&
-        inter_vehicle_distance > minimum_inter_vehicle_distance) &&
-
-        !(!reference_front_wp->CheckJunction() &&
-        cg::Math::Dot(reference_heading, reference_to_other) > 0 &&
-        inter_vehicle_distance > std::max(minimum_inter_vehicle_distance,
-                                          std::pow(parameters.GetDistanceToLeadingVehicle(reference_vehicle)
-                                                   + inter_vehicle_length, 2.0f)))) {
-
+    float ego_bounding_box_extension = GetBoundingBoxExtention(reference_vehicle);
+    float other_bounding_box_extension = GetBoundingBoxExtention(other_vehicle);
+    // Calculate minimum distance between vehicle to consider collision negotiation.
+    float inter_vehicle_length = reference_vehicle_length + other_vehicle_length;
+    float ego_detection_range = std::pow(ego_bounding_box_extension + inter_vehicle_length, 2.0f);
+    float cross_detection_range = std::pow(ego_bounding_box_extension + inter_vehicle_length
+                                           + other_bounding_box_extension, 2.0f);
+
+    // Conditions to consider collision negotiation.
+    bool other_vehicle_in_ego_range = inter_vehicle_distance < ego_detection_range;
+    bool other_vehicles_in_cross_detection_range = inter_vehicle_distance < cross_detection_range;
+    bool other_vehicle_in_front = cg::Math::Dot(reference_heading, reference_to_other) > 0;
+    bool ego_inside_junction = closest_point->CheckJunction();
+    bool ego_at_traffic_light = reference_vehicle_ptr->IsAtTrafficLight();
+    bool ego_stopped_by_light = reference_vehicle_ptr->GetTrafficLightState() != carla::rpc::TrafficLightState::Green;
+    bool ego_at_junction_entrance = !closest_point->CheckJunction() && junction_look_ahead->CheckJunction();
+
+    const ActorId reference_vehicle_id = reference_vehicle->GetId();
+    // Conditions to avoid collision negotiation.
+    if (!(ego_at_junction_entrance && ego_at_traffic_light && ego_stopped_by_light)
+        && !(!ego_inside_junction && !other_vehicle_in_front)
+        && !(!ego_inside_junction && other_vehicle_in_front && !other_vehicle_in_ego_range)
+        && !(ego_inside_junction && !other_vehicles_in_cross_detection_range)
+    ) {
```

Comment:
Instead one can use below condition which uses 10 operations whereas above one uses 19 operations. 
    if (!(ego_at_junction_entrance && ego_at_traffic_light && ego_stopped_by_light)
        && ((ego_inside_junction && other_vehicles_in_cross_detection_range)
                || (other_vehicle_in_front && other_vehicle_in_ego_range 
                    && (!ego_inside_junction || other_vehicles_in_cross_detection_range))
    )

Let,
ego_at_junction_entrance = a 
ego_at_traffic_light = b
ego_stopped_by_light = c
ego_inside_junction = d
other_vehicle_in_front = e
other_vehicle_in_ego_range = f
other_vehicles_in_cross_detection_range = g

Original Boolean condition can be simplified as,
    (abc)' (d'e')' (d'ef')' (dg')' 
=  (abc)' (d + e) (d + e' + f) (d' + g)                                                               [DeMorgan's Law]
=  (abc)' (d + e' + f) (d + e) (d' + g)                                                               [Commutative Law]      
=  (abc)' (d + e' + f)  (dd' + dg + d'e + eg)                                                     [Distributive Law]
=  (abc)' (d + e' + f)  (dg + d'e + eg)                                                              [Complement Law: AA' = F]
=  (abc)' (ddg + dd'e + deg + de'g + d'ee' + ee'f + dfg + d'ef + efg)               [Distributive Law]
=  (abc)' (dg +              deg + de'g +                   + dfg + d'ef + efg)               [Complement Law: AA' = F]
=  (abc)' (dg +              dg(e + e')   +                   + dfg + d'ef + efg)               [Distributive Law]
=  (abc)' (dg +              dg +                                + dfg + d'ef + efg)               [Complement Law: A+A' = 1]
=  (abc)' (dg +              dg ( 1 + f)                                 + d'ef + efg)               [Distributive Law]
=  (abc)' (dg +              dg +                                         + d'ef + efg)                [Identity Law: 1+A = 1, A.1 = A]
=  (abc)' (dg + d'ef + efg)                                                                                [Idempotent Law AA = A]
=  (abc)' (dg + ef(d' + g))  

                                                                               
[truth-table.xlsx](https://github.com/carla-simulator/carla/files/4384030/truth-table.xlsx)